### PR TITLE
Added review of a paper: Generative Adversarial Networks, Goodfellow et al, 2014.

### PR DESCRIPTION
**Pull request by Monnoroch**

### CODE REVIEW

File: magenta/reviews/GAN.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+<p align="center">
+  <img src="assets/gan/image00.gif"><br>
+  <i>Here are a set of faces generated purely from random noise.</i><br>
+  <i>Source: http://torch.ch/blog/2015/11/13/gan.html.</i>
+</p>
+
+So far deep generative models have had much less success than deep discriminative models.
+This is partly due to difficulty in approximating intractable probabilistic posteriors that arise in their
+optimization algorithms, which forces practitioners to use techniques like MCMC.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Write out the full name and link it to an explanation.

File: magenta/reviews/GAN.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+It’s also because many of the methods that help train discriminative models do not work as well when applied to
+generative models.
+The adversarial framework in this paper can be trained purely with back propagation and helps solve some of
+these problems.
+GANs are an idea rooted in Game Theory.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: New paragraph.

File: magenta/reviews/GAN.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+Two models compete against each other.
+The first, the generator, generates examples that it thinks are real.
+The second, the discriminator, tries to discriminate between examples from the dataset and examples drawn
+from the pool generated by the generator.
+The discriminator is trained in a supervised manner with two labels -- real or generated.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "-- real or generated" --> "- Real and Generated"

File: magenta/reviews/GAN.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+Those familiar with game theory might appreciate the insight that the optimization of this objective is like finding
+a Nash equilibrium of a minimax game between the generator and the discriminator.
+The loss has two components.
+The first one, <img src="assets/gan/image06.png">, makes the discriminator <img src="assets/gan/image01.png"> better at
+discriminating real examples from fake ones, and the second one, <img src="assets/gan/image13.png">, makes the generator
+<img src="assets/gan/image14.png"> generate samples that the discriminator considers real.
+The input, <img src="assets/gan/image02.png">, is a noise vector with distribution <img src="assets/gan/image10.png">.
+This is usually an N-dimensional normal or uniform distribution and serves as input to the network.
+
+Both parts can be independently optimized with gradient based optimization in turns.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "independently optimized with gradient based optimization in turns" --> "independently optimized with gradient based methods".

File: magenta/reviews/GAN.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+The generator maps this <img src="assets/gan/image02.png"> to <img src="assets/gan/image04.png"> which results in fake examples,
+distributed according to the green line on the figure.
+The dashed blue line is the discriminator and gives each example from <img src="assets/gan/image04.png"> a probability of
+being real.
+
+Starting with a fixed, random generator, and a discriminator (see Figure (a)), we begin by training the discriminator
+to distinguish real examples from fake examples produced by the generator (see Figure (b)).
+Once trained, we fix the discriminator and train the generator to maximize discriminator’s error (Figure (c)).
+Then we train the discriminator again, and so forth, until convergence.
+When it converges, the generator had learned a true data distribution which matches the black dotted line exactly.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: s/had/has

File: magenta/reviews/GAN.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+Therefore the discriminator has no way of telling whether an example is real or fake, it returns ½ on all of them
+(Figure (d)).
+
+Under certain conditions, this process converges to a fixed point, where the generator has learned the true
+data distribution, and hence the discriminator cannot classify real examples from generated ones.
+In practice, the original paper’s model is capable of generating great results on simple tasks such as MNIST.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Remove "In practice,".

File: magenta/reviews/GAN.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+Recent advances have made it possible to get higher quality more complicated problems like generating faces and CIFAR
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Make this sentence readable.

File: magenta/reviews/GAN.md [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+images.
+
+## Experiments.
+
+<p align="center">
+ <img src="assets/gan/image05.png"><br>
+</p>
+
+To evaluate model performance quantitatively one could calculate the likelihood of the real examples given the
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "," after quantitatively.

File: magenta/reviews/GAN.md [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+distribution that the model induces.
+On MNIST examples, generated by GANs give approximated log likelihood 225 with standard error 2 across several
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Grammar.

File: magenta/reviews/GAN.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+dataset folds.
+In the same setting Deep Belief Network, which is an undirected graphical model, gives approximated log likelihood 138
+with similar error.
+Examples generated from Toronto Face Dataset, which is a much harder problem, gives approximated log likelihood of
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Same.

File: magenta/reviews/GAN.md (hunk 2 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 3 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 4 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 5 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 6 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 8 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 9 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 10 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 7 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 1 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 3 above)
Comment (cinjon) [inline magenta/reviews/GAN.md]: "--" --> "-"

File: magenta/reviews/GAN.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+The first one, <img src="assets/gan/image06.png">, makes the discriminator <img src="assets/gan/image01.png"> better at
+discriminating real examples from fake ones, and the second one, <img src="assets/gan/image13.png">, makes the generator
+<img src="assets/gan/image14.png"> generate samples that the discriminator considers real.
+The input, <img src="assets/gan/image02.png">, is a noise vector with distribution <img src="assets/gan/image10.png">.
+This is usually an N-dimensional normal or uniform distribution and serves as input to the network.
+
+Both parts can be independently optimized with gradient based methods.
+First, do an optimization step for the discriminator to maximize <img src="assets/gan/image08.png">, then do
+an optimization step for the generator to make it better at fooling this new discriminator by minimizing
+<img src="assets/gan/image12.png">. If <img src="assets/gan/image03.png"> is very low, which is the case early in training,
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "... in training, ..." --> "... in training, then ..."

File: magenta/reviews/GAN.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+Here’s a visual explanation of how the learning happens.
+
+<p align="center">
+ <img src="assets/gan/image09.png"><br>
+ <i>Training a GAN to fit a data distribution.</i>
+</p>
+
+Imagine that the real data is represented by some distribution, drawn as a dotted black line on the figure.
+Real examples then are obtained by sampling from it.
+<img src="assets/gan/image04.png"> is the domain of the model's input.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: " is the domain of the model's input." --> "The domain of the model's input is ." It's odd seeing the lowercase x in the beginning.

File: magenta/reviews/GAN.md [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+Fake examples are obtained by sampling from some distribution on domain <img src="assets/gan/image02.png">.
+The generator maps this <img src="assets/gan/image02.png"> to <img src="assets/gan/image04.png"> which results in fake examples,
+whicha are distributed according to the green line on the figure.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Fix this sentence grammar please. Also, note the "whicha".

File: magenta/reviews/GAN.md [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+The dashed blue line is the discriminator and it gives each example from <img src="assets/gan/image04.png"> a probability of
+being real.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "... of being from the real data distribution."

File: magenta/reviews/GAN.md [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+Starting with a fixed, random generator and a discriminator (see Figure (a)), we begin by training the discriminator
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "Starting with a fixed, random generator and a discriminator". Please fix this grammar. If I didn't know this work, I would not understand this sentence. What you mean to say is that the models start from a fixed but random initialization. Then say that. Be like Hemingway.

File: magenta/reviews/GAN.md [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+to distinguish real examples from fake examples produced by the generator (see Figure (b)).
+Once trained, we fix the discriminator and train the generator to maximize discriminator’s error (Figure (c)).
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "maximize THE discriminator's..."

File: magenta/reviews/GAN.md [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+Then we train the discriminator again, and so forth, until convergence.
+When it converges, the generator has learned a true data distribution which matches the black dotted line exactly.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "When it converges ..." --> "At this point ..."

File: magenta/reviews/GAN.md [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+Therefore the discriminator has no way of telling whether an example is real or fake, it returns ½ on all of them
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Grammar.

File: magenta/reviews/GAN.md [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+## Experiments.
+
+<p align="center">
+ <img src="assets/gan/image05.png"><br>
+</p>
+
+To evaluate model performance quantitatively, one could calculate the likelihood of the real examples given the
+distribution that the model induces.
+MNIST examples generated by GANs have approximated log likelihood 225 with standard error 2 across several
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: I don't understand this sentence.

File: magenta/reviews/GAN.md [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+dataset folds.
+In the same setting Deep Belief Network, which is an undirected graphical model, generates examples with approximated log likelihood 138
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Grammar.

File: magenta/reviews/GAN.md [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+with similar error.
+Examples generated from Toronto Face Dataset, which is a much harder problem, have approximated log likelihood of
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Do you mean to say "approximately"? I'm honestly not sure.

File: magenta/reviews/GAN.md [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+2057 ± 26 versus 1909 ± 66 with DBN.
+
+## Disadvantages
+GANs are difficult to optimize.
+The two networks have to be kept “in sync”.
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: Get rid of the quotations.

File: magenta/reviews/GAN.md [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+If the discriminator wins by too big a margin, then the generator can’t learn as the discriminator error is too small.
+However, if the generator wins by too much, it will have trouble learning because discriminator is too weak
```
Comment (cinjon) [inline magenta/reviews/GAN.md]: "the discriminator"

File: magenta/reviews/GAN.md (hunk 11 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 12 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 13 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 14 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 16 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 17 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 23 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 22 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: Done.

File: magenta/reviews/GAN.md (hunk 21 above)
Comment (Monnoroch) [inline magenta/reviews/GAN.md]: I mean that we cannot calculate the true log likelihood, so we approximate in somehow.

**Comment by cinjon** [comment]
LGTM.

**Pull request merged**